dcicutils/glacier_utils.py

```diff
@@ -0,0 +1,455 @@
+                                restore_length: int = 7, new_status: str = 'uploaded', storage_class: str = 'STANDARD',
+                                parallel: bool = False, num_threads: int = 4, delete_all_versions: bool = False,
+                                phase: int = 1) -> (List[str], List[str]):
+        """ Overarching method that will take a search query and loop through all files in the
+            search results, running the appropriate phase as passed
+
+        :param search_query: search query used to resolve items desired to be restored
+        :param page_limit: number of pages to resolve ie: 25 * page_limit items, 25 * 50 = 1250 items by default
+        :param search_generator: whether to use a generator - can be useful in some steps, but not in the copy
+        :param restore_length: length of time for restore to be active in days
+        :param new_status: status to patch to file items
+        :param storage_class: new storage class for copy
+        :param parallel: whether to use the parallel copy
+        :param num_threads: number of threads to use if parallel is active
+        :param delete_all_versions: if deleting, whether to clear ALL glacier versions
+        :param phase: which phase of the glacier restore to run, one of [1, 2, 3, 4]
+        :return: 2-tuple of successful, failed @ids extracted from search
+        """
+        if phase not in [1, 2, 3, 4]:
+            raise GlacierRestoreException(f'Invalid phase passed to restore_all_from_search: {phase},'
+                                          f' valid phases: [1, 2, 3, 4]\n'
+                                          f'Phase 1: Issue Restore\n'
+                                          f'Phase 2: Issue Copy\n'
+                                          f'Phase 3: Patch lifecycle status\n'
+                                          f'Phase 4: Delete Glacier copies')
+        success, failed = [], []
+        if search_generator:  # operate on search results individually, fine for all phases except the copy (2)
+            for item_meta in search_metadata(
+                search_query,
+                key=self.env_key, ff_env=self.env_name,
+                page_limit=page_limit, is_generator=search_generator
+            ):
+                atid = item_meta['@id']
+                if phase == 1:
+                    current_success, current_failed = self.restore_portal_from_glacier(atid, item_meta,
+                                                                                       days=restore_length)
+                    if current_failed:
+                        failed.append(atid)
+                    else:
+                        success.append(atid)
+                elif phase == 2:
+                    raise GlacierRestoreException(f'Invalid phase for search_generator=True!'
+                                                  f' Do not use a generator when doing the copy phase.')
```

Comment:
Is this only a bad idea when `parallel` is true? Seems like if not running in parallel there's no big difference between passing in the individual search results vs. all of them together.